source/tools/MatchBin.h

```diff
@@ -23,15 +28,288 @@
 #include <math.h>
 #include <mutex>
 #include <shared_mutex>
+#include <string>
+#include <unordered_map>
+#include <atomic>
 
 #include "../base/assert.h"
 #include "../base/vector.h"
 #include "../tools/IndexMap.h"
 #include "../tools/BitSet.h"
 #include "../tools/matchbin_utils.h"
+#include "../data/DataFile.h"
+#include "../data/DataNode.h"
+#include "../base/errors.h"
 
 namespace emp {
+  template <
+    typename Val,
+    typename Metric,
+    typename Selector,
+    typename Regulator
+  > class MatchBin;
+}
+
+// namespace used for logging and caching
+namespace emp::internal {
+  template<typename Query, typename Tag>
+  class MatchBinLog {
+    using query_t = Query;
+    using tag_t = Tag;
+
+    template <
+      typename Val,
+      typename Metric,
+      typename Selector,
+      typename Regulator
+    >
+    friend class emp::MatchBin;
+
+
+    struct LogEntry {
+      query_t query;
+      std::optional<tag_t> maybe_tag;
+      std::string buffer;
+
+      operator size_t() const { return emp::CombineHash(query, maybe_tag, buffer); }
+    };
+
+    using logbuffer_t = std::unordered_map<
+      LogEntry,
+      size_t,
+      TupleHash<
+        LogEntry
+      >
+    >;
+
+    size_t log_counter;
+    size_t id;
+    logbuffer_t logbuffer;
+
+    #ifdef EMP_LOG_MATCHBIN
+    static constexpr bool logging_enabled = true;
+    #else
+    static constexpr bool logging_enabled = false;
+    #endif
+    bool logging_activated{logging_enabled};
```

Comment:
On further reflection, we should probably always initialize this to false and then have the user enable it

Reply:
Done in 41cfb50 and d0b9d53.